### ISSUE TITLE
Configurable SSH options for async fetching

### ISSUE THREAD

**Issue by madduck** [munin-monitoring/munin#694]
For the `ssh` URI scheme, munin hard-codes SSH options, including `StrictHostKeyChecking=no`. While I think that this is prohibitive, it's probably the path of least resistance and might well be required in certain environments.

However, in many situations, specifying `StrictHostKeyChecking=no` degrades the security.

At the same time, other SSH client options could be useful in other situations, for instance compression, proxy commands, identities file, and IPQoS, to just name a few.

Long story short, please make the SSH options configurable. There could be a default, but then it'd be good to specify site-global options (i.e. override the default), and the ability to add options per-host.

Thanks,
-martin

**Comment by ssm** [comment]
I've made two commits, one for "stable-2.0" and one for the "devel" branch. They are linked from this bug report, and will soon be pull requests.

If you would like to test, please apply the relevant lines on the master.  (It works here, but that's no guarantee :)

Example config:

    # /etc/munin/munin-conf.d/config.ssh
    ssh_options -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o PreferredAuthentications=publickey

**Comment by madduck** [comment]
Hi, amazing how quickly you went for this, and I am sorry that I am only replying now. I am going to try this out very soon.

How hard would it be now to allow per-host overriding of these options?

**Comment by ssm** [comment]
It does not look to be very hard.

I _think_ the node definition config is available as $self->{configref} at that location in the code, so $self->{configref}->{ssh_options} could, if present, override $config->{ssh_options} which would be the global configuration item.

That would mean, for parsing a hypothetical munin.conf, at https://github.com/ssm/munin/blob/694-ssh-option-stable-2.0/master/lib/Munin/Master/Node.pm#L96, the relevant variables would be :

    # global options
    ssh_options "blah blah"        # $config->{ssh_options}

    # node definitions
    [node.example.com]
      ssh_options "more blah blah" # $self->{configref}->{ssh_options}

The node definition would then override the global setting, which would override the default.

Looks sane?

**Comment by madduck** [comment]
Yeah, that'd be exactly the expected usage. Thanks so much for your quick turnaround!

**Comment by shallot** [comment]
It should just be noted that doing

```
ssh_options "-o foo -o bar"
```

...actually breaks everything :) the quotes need to go, 

```
ssh_options -o foo -o bar
```

**Comment by sumpfralle** [comment]
Good to know!

Do you think, that we should clarify this somewhere in the [documentation](http://guide.munin-monitoring.org/en/latest/reference/munin.conf.html#cmdoption-munin-conf-arg-ssh-options) or somewhere else?

**Comment by paulgevers** [comment]
Several of my munin nodes are behind the same JumpHost and we need to connect to them via different ssh ports. However, it seems that the `Port 12345` part of the ssh_config is ignored as in debugging mode I see port 22. When I add the port to the hosts.conf it works, but it took me quite some time to figure it out, because logged in as `munin` user on the main node I could correctly ssh into the munin node, so things seemed to be correctly set upped. Is that worth a bug report?
